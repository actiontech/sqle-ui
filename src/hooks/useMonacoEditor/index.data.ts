import monacoEditor from 'monaco-editor';
import { IRange } from './index.type';

export function createDependencyProposals(
  monaco: typeof monacoEditor,
  range: IRange
) {
  const sqlKeyWords = [
    'ADD',
    'ADD CONSTRAINT',
    'ALTER',
    'ALTER COLUMN',
    'ALTER TABLE',
    'ALL',
    'AND',
    'ANY',
    'AS',
    'ASC',
    'BACKUP DATABASE',
    'BETWEEN',
    'CASE',
    'CHECK',
    'COLUMN',
    'CONSTRAINT',
    'CREATE',
    'CREATE DATABASE',
    'CREATE INDEX',
    'CREATE OR REPLACE VIEW',
    'CREATE TABLE',
    'CREATE PROCEDURE',
    'CREATE UNIQUE INDEX',
    'CREATE VIEW',
    'DATABASE',
    'DEFAULT',
    'DELETE',
    'DESC',
    'DISTINCT',
    'DROP',
    'DROP COLUMN',
    'DROP CONSTRAINT',
    'DROP DATABASE',
    'DROP DEFAULT',
    'DROP INDEX',
    'DROP TABLE',
    'DROP VIEW',
    'EXEC',
    'EXISTS',
    'FROM',
    'FOREIGN KEY',
    'FULL OUTER JOIN',
    'GROUP BY',
    'HAVING',
    'IN',
    'INDEX',
    'INNER JOIN',
    'INSERT INTO',
    'INSERT INTO SELECT',
    'IS NULL',
    'IS NOT NULL',
    'JOIN',
    'LEFT JOIN',
    'LIKE',
    'LIMIT',
    'NOT',
    'NOT NULL',
    'OR',
    'ORDER BY',
    'OUTER JOIN',
    'PRIMARY KEY',
    'PROCEDURE',
    'RIGHT JOIN',
    'ROWNUM',
    'SELECT',
    'SELECT DISTINCT',
    'SELECT INTO',
    'SELECT TOP',
    'SET',
    'TABLE',
    'TOP',
    'TRUNCATE TABLE',
    'UNION',
    'UNION ALL',
    'UNIQUE',
    'UPDATE',
    'VALUES',
    'VIEW',
    'WHERE',
  ];

  return sqlKeyWords.map((v) => ({
    label: v,
    kind: monaco.languages.CompletionItemKind.Function,
    insertText: v,
    range,
  }));
}
