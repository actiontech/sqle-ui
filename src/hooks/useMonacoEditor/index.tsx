import { FormInstance } from 'antd';
import { EditorDidMount } from 'react-monaco-editor';

const useMonacoEditor = (
  form?: FormInstance,
  {
    formName,
    placeholder = '/* input your sql */',
  }: { formName?: string; placeholder?: string } = {}
) => {
  const editorDidMount: EditorDidMount = (editor) => {
    editor.onDidFocusEditorText(() => {
      if (!!form && !!formName) {
        const currentValue = form.getFieldValue(formName);
        if (currentValue === placeholder) {
          form.setFieldsValue({
            [formName]: '',
          });
        }
      } else {
        const currentValue = editor.getValue();
        if (currentValue === placeholder) {
          editor.setValue('');
        }
      }
    });

    editor.onDidBlurEditorText(() => {
      if (!!form && !!formName) {
        const currentValue = form.getFieldValue(formName);
        if (currentValue === '') {
          form.setFieldsValue({
            [formName]: placeholder,
          });
        }
      } else {
        const currentValue = editor.getValue();
        if (currentValue === '') {
          editor.setValue(placeholder);
        }
      }
    });
  };

  return {
    editorDidMount,
  };
};

export default useMonacoEditor;
