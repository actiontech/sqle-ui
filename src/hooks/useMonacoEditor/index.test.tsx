import { renderHook } from '@testing-library/react-hooks';
import useMonacoEditor from '.';

describe('userMonacoEditor', () => {
  test('should set value to empty when editor value is equal default placeholder', () => {
    const editor = {
      onDidFocusEditorText: jest.fn(),
      onDidBlurEditorText: jest.fn(),
      getValue: jest.fn().mockReturnValue('getValue'),
      setValue: jest.fn(),
    };
    const { result } = renderHook(() => useMonacoEditor());
    result.current.editorDidMount(editor as any, {} as any);

    expect(editor.onDidBlurEditorText).toBeCalledTimes(1);
    expect(editor.onDidFocusEditorText).toBeCalledTimes(1);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    const focusFn = editor.onDidFocusEditorText.mock.calls[0][0];
    focusFn();
    expect(editor.getValue).toBeCalledTimes(1);
    expect(editor.setValue).toBeCalledTimes(0);

    editor.getValue.mockReturnValue('/* input your sql */');

    focusFn();
    expect(editor.getValue).toBeCalledTimes(2);
    expect(editor.setValue).toBeCalledTimes(1);
    expect(editor.setValue).toBeCalledWith('');
  });

  test('should set value to empty when editor value is equal custom placeholder', () => {
    const editor = {
      onDidFocusEditorText: jest.fn(),
      onDidBlurEditorText: jest.fn(),
      getValue: jest.fn().mockReturnValue('getValue'),
      setValue: jest.fn(),
    };
    const { result } = renderHook(() =>
      useMonacoEditor(undefined, { placeholder: 'test aaaa' })
    );
    result.current.editorDidMount(editor as any, {} as any);

    expect(editor.onDidBlurEditorText).toBeCalledTimes(1);
    expect(editor.onDidFocusEditorText).toBeCalledTimes(1);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    const focusFn = editor.onDidFocusEditorText.mock.calls[0][0];
    editor.getValue.mockReturnValue('/* input your sql */');
    focusFn();
    expect(editor.getValue).toBeCalledTimes(1);
    expect(editor.setValue).toBeCalledTimes(0);

    editor.getValue.mockReturnValue('test aaaa');
    focusFn();
    expect(editor.getValue).toBeCalledTimes(2);
    expect(editor.setValue).toBeCalledTimes(1);
    expect(editor.setValue).toBeCalledWith('');
  });

  test('should set value to default placeholder when editor is blur and value is empty', () => {
    const editor = {
      onDidFocusEditorText: jest.fn(),
      onDidBlurEditorText: jest.fn(),
      getValue: jest.fn().mockReturnValue('getValue'),
      setValue: jest.fn(),
    };
    const { result } = renderHook(() => useMonacoEditor());
    result.current.editorDidMount(editor as any, {} as any);

    expect(editor.onDidBlurEditorText).toBeCalledTimes(1);
    expect(editor.onDidFocusEditorText).toBeCalledTimes(1);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    const blurFn = editor.onDidBlurEditorText.mock.calls[0][0];

    blurFn();
    expect(editor.getValue).toBeCalledTimes(1);
    expect(editor.setValue).toBeCalledTimes(0);

    editor.getValue.mockReturnValue('');

    blurFn();
    expect(editor.getValue).toBeCalledTimes(2);
    expect(editor.setValue).toBeCalledTimes(1);
    expect(editor.setValue).toBeCalledWith('/* input your sql */');
  });

  test('should set value to custom placeholder when editor is blur and value is empty', () => {
    const editor = {
      onDidFocusEditorText: jest.fn(),
      onDidBlurEditorText: jest.fn(),
      getValue: jest.fn().mockReturnValue('getValue'),
      setValue: jest.fn(),
    };
    const { result } = renderHook(() =>
      useMonacoEditor(undefined, { placeholder: 'test aaa' })
    );
    result.current.editorDidMount(editor as any, {} as any);

    expect(editor.onDidBlurEditorText).toBeCalledTimes(1);
    expect(editor.onDidFocusEditorText).toBeCalledTimes(1);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    const blurFn = editor.onDidBlurEditorText.mock.calls[0][0];

    blurFn();
    expect(editor.getValue).toBeCalledTimes(1);
    expect(editor.setValue).toBeCalledTimes(0);

    editor.getValue.mockReturnValue('');

    blurFn();
    expect(editor.getValue).toBeCalledTimes(2);
    expect(editor.setValue).toBeCalledTimes(1);
    expect(editor.setValue).toBeCalledWith('test aaa');
  });

  test('should set value to empty by form instance when editor value is equal default placeholder', () => {
    const editor = {
      onDidFocusEditorText: jest.fn(),
      onDidBlurEditorText: jest.fn(),
      getValue: jest.fn().mockReturnValue('getValue'),
      setValue: jest.fn(),
    };
    const form = {
      getFieldValue: jest.fn(),
      setFieldsValue: jest.fn(),
    };
    const { result } = renderHook(() =>
      useMonacoEditor(form as any, { formName: 'sql' })
    );
    result.current.editorDidMount(editor as any, {} as any);

    expect(editor.onDidBlurEditorText).toBeCalledTimes(1);
    expect(editor.onDidFocusEditorText).toBeCalledTimes(1);

    const focusFn = editor.onDidFocusEditorText.mock.calls[0][0];
    focusFn();
    expect(form.getFieldValue).toBeCalledTimes(1);
    expect(form.getFieldValue).toBeCalledWith('sql');
    expect(form.setFieldsValue).toBeCalledTimes(0);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    form.getFieldValue.mockReturnValue('/* input your sql */');

    focusFn();
    expect(form.getFieldValue).toBeCalledTimes(2);
    expect(form.getFieldValue).lastCalledWith('sql');
    expect(form.setFieldsValue).toBeCalledTimes(1);
    expect(form.setFieldsValue).toBeCalledWith({ sql: '' });
  });

  test('should set value to empty by form instance when editor value is equal custom placeholder', () => {
    const editor = {
      onDidFocusEditorText: jest.fn(),
      onDidBlurEditorText: jest.fn(),
      getValue: jest.fn().mockReturnValue('getValue'),
      setValue: jest.fn(),
    };
    const form = {
      getFieldValue: jest.fn(),
      setFieldsValue: jest.fn(),
    };
    const { result } = renderHook(() =>
      useMonacoEditor(form as any, {
        formName: 'sql',
        placeholder: 'test aaaa',
      })
    );
    result.current.editorDidMount(editor as any, {} as any);

    expect(editor.onDidBlurEditorText).toBeCalledTimes(1);
    expect(editor.onDidFocusEditorText).toBeCalledTimes(1);

    const focusFn = editor.onDidFocusEditorText.mock.calls[0][0];
    focusFn();
    expect(form.getFieldValue).toBeCalledTimes(1);
    expect(form.getFieldValue).toBeCalledWith('sql');
    expect(form.setFieldsValue).toBeCalledTimes(0);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    form.getFieldValue.mockReturnValue('test aaaa');

    focusFn();
    expect(form.getFieldValue).toBeCalledTimes(2);
    expect(form.getFieldValue).lastCalledWith('sql');
    expect(form.setFieldsValue).toBeCalledTimes(1);
    expect(form.setFieldsValue).toBeCalledWith({ sql: '' });
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);
  });

  test('should set value to default placeholder by form instance when editor is blur and value is empty', () => {
    const editor = {
      onDidFocusEditorText: jest.fn(),
      onDidBlurEditorText: jest.fn(),
      getValue: jest.fn().mockReturnValue('getValue'),
      setValue: jest.fn(),
    };
    const form = {
      getFieldValue: jest.fn(),
      setFieldsValue: jest.fn(),
    };
    const { result } = renderHook(() =>
      useMonacoEditor(form as any, { formName: 'sql' })
    );
    result.current.editorDidMount(editor as any, {} as any);

    expect(editor.onDidBlurEditorText).toBeCalledTimes(1);
    expect(editor.onDidFocusEditorText).toBeCalledTimes(1);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    const blurFn = editor.onDidBlurEditorText.mock.calls[0][0];

    blurFn();
    expect(form.getFieldValue).toBeCalledTimes(1);
    expect(form.getFieldValue).toBeCalledWith('sql');
    expect(form.setFieldsValue).toBeCalledTimes(0);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    form.getFieldValue.mockReturnValue('');

    blurFn();
    expect(form.getFieldValue).toBeCalledTimes(2);
    expect(form.getFieldValue).lastCalledWith('sql');
    expect(form.setFieldsValue).toBeCalledTimes(1);
    expect(form.setFieldsValue).toBeCalledWith({ sql: '/* input your sql */' });
  });

  test('should set value to custom placeholder by form instance when editor is blur and value is empty', () => {
    const editor = {
      onDidFocusEditorText: jest.fn(),
      onDidBlurEditorText: jest.fn(),
      getValue: jest.fn().mockReturnValue('getValue'),
      setValue: jest.fn(),
    };
    const form = {
      getFieldValue: jest.fn(),
      setFieldsValue: jest.fn(),
    };
    const { result } = renderHook(() =>
      useMonacoEditor(form as any, {
        formName: 'sql',
        placeholder: 'test aaaa',
      })
    );
    result.current.editorDidMount(editor as any, {} as any);

    expect(editor.onDidBlurEditorText).toBeCalledTimes(1);
    expect(editor.onDidFocusEditorText).toBeCalledTimes(1);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    const blurFn = editor.onDidBlurEditorText.mock.calls[0][0];

    blurFn();
    expect(form.getFieldValue).toBeCalledTimes(1);
    expect(form.getFieldValue).toBeCalledWith('sql');
    expect(form.setFieldsValue).toBeCalledTimes(0);
    expect(editor.getValue).toBeCalledTimes(0);
    expect(editor.setValue).toBeCalledTimes(0);

    form.getFieldValue.mockReturnValue('');

    blurFn();
    expect(form.getFieldValue).toBeCalledTimes(2);
    expect(form.getFieldValue).lastCalledWith('sql');
    expect(form.setFieldsValue).toBeCalledTimes(1);
    expect(form.setFieldsValue).toBeCalledWith({ sql: 'test aaaa' });
  });
});
