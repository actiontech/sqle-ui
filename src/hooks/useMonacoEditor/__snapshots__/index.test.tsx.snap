// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userMonacoEditor should match snapshot for createDependencyProposals 1`] = `
Array [
  Object {
    "insertText": "ADD",
    "kind": "function",
    "label": "ADD",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ADD CONSTRAINT",
    "kind": "function",
    "label": "ADD CONSTRAINT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ALTER",
    "kind": "function",
    "label": "ALTER",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ALTER COLUMN",
    "kind": "function",
    "label": "ALTER COLUMN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ALTER TABLE",
    "kind": "function",
    "label": "ALTER TABLE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ALL",
    "kind": "function",
    "label": "ALL",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "AND",
    "kind": "function",
    "label": "AND",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ANY",
    "kind": "function",
    "label": "ANY",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "AS",
    "kind": "function",
    "label": "AS",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ASC",
    "kind": "function",
    "label": "ASC",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "BACKUP DATABASE",
    "kind": "function",
    "label": "BACKUP DATABASE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "BETWEEN",
    "kind": "function",
    "label": "BETWEEN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CASE",
    "kind": "function",
    "label": "CASE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CHECK",
    "kind": "function",
    "label": "CHECK",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "COLUMN",
    "kind": "function",
    "label": "COLUMN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CONSTRAINT",
    "kind": "function",
    "label": "CONSTRAINT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CREATE",
    "kind": "function",
    "label": "CREATE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CREATE DATABASE",
    "kind": "function",
    "label": "CREATE DATABASE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CREATE INDEX",
    "kind": "function",
    "label": "CREATE INDEX",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CREATE OR REPLACE VIEW",
    "kind": "function",
    "label": "CREATE OR REPLACE VIEW",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CREATE TABLE",
    "kind": "function",
    "label": "CREATE TABLE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CREATE PROCEDURE",
    "kind": "function",
    "label": "CREATE PROCEDURE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CREATE UNIQUE INDEX",
    "kind": "function",
    "label": "CREATE UNIQUE INDEX",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "CREATE VIEW",
    "kind": "function",
    "label": "CREATE VIEW",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DATABASE",
    "kind": "function",
    "label": "DATABASE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DEFAULT",
    "kind": "function",
    "label": "DEFAULT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DELETE",
    "kind": "function",
    "label": "DELETE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DESC",
    "kind": "function",
    "label": "DESC",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DISTINCT",
    "kind": "function",
    "label": "DISTINCT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DROP",
    "kind": "function",
    "label": "DROP",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DROP COLUMN",
    "kind": "function",
    "label": "DROP COLUMN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DROP CONSTRAINT",
    "kind": "function",
    "label": "DROP CONSTRAINT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DROP DATABASE",
    "kind": "function",
    "label": "DROP DATABASE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DROP DEFAULT",
    "kind": "function",
    "label": "DROP DEFAULT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DROP INDEX",
    "kind": "function",
    "label": "DROP INDEX",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DROP TABLE",
    "kind": "function",
    "label": "DROP TABLE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "DROP VIEW",
    "kind": "function",
    "label": "DROP VIEW",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "EXEC",
    "kind": "function",
    "label": "EXEC",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "EXISTS",
    "kind": "function",
    "label": "EXISTS",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "FROM",
    "kind": "function",
    "label": "FROM",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "FOREIGN KEY",
    "kind": "function",
    "label": "FOREIGN KEY",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "FULL OUTER JOIN",
    "kind": "function",
    "label": "FULL OUTER JOIN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "GROUP BY",
    "kind": "function",
    "label": "GROUP BY",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "HAVING",
    "kind": "function",
    "label": "HAVING",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "IN",
    "kind": "function",
    "label": "IN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "INDEX",
    "kind": "function",
    "label": "INDEX",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "INNER JOIN",
    "kind": "function",
    "label": "INNER JOIN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "INSERT INTO",
    "kind": "function",
    "label": "INSERT INTO",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "INSERT INTO SELECT",
    "kind": "function",
    "label": "INSERT INTO SELECT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "IS NULL",
    "kind": "function",
    "label": "IS NULL",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "IS NOT NULL",
    "kind": "function",
    "label": "IS NOT NULL",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "JOIN",
    "kind": "function",
    "label": "JOIN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "LEFT JOIN",
    "kind": "function",
    "label": "LEFT JOIN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "LIKE",
    "kind": "function",
    "label": "LIKE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "LIMIT",
    "kind": "function",
    "label": "LIMIT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "NOT",
    "kind": "function",
    "label": "NOT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "NOT NULL",
    "kind": "function",
    "label": "NOT NULL",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "OR",
    "kind": "function",
    "label": "OR",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ORDER BY",
    "kind": "function",
    "label": "ORDER BY",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "OUTER JOIN",
    "kind": "function",
    "label": "OUTER JOIN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "PRIMARY KEY",
    "kind": "function",
    "label": "PRIMARY KEY",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "PROCEDURE",
    "kind": "function",
    "label": "PROCEDURE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "RIGHT JOIN",
    "kind": "function",
    "label": "RIGHT JOIN",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "ROWNUM",
    "kind": "function",
    "label": "ROWNUM",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "SELECT",
    "kind": "function",
    "label": "SELECT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "SELECT DISTINCT",
    "kind": "function",
    "label": "SELECT DISTINCT",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "SELECT INTO",
    "kind": "function",
    "label": "SELECT INTO",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "SELECT TOP",
    "kind": "function",
    "label": "SELECT TOP",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "SET",
    "kind": "function",
    "label": "SET",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "TABLE",
    "kind": "function",
    "label": "TABLE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "TOP",
    "kind": "function",
    "label": "TOP",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "TRUNCATE TABLE",
    "kind": "function",
    "label": "TRUNCATE TABLE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "UNION",
    "kind": "function",
    "label": "UNION",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "UNION ALL",
    "kind": "function",
    "label": "UNION ALL",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "UNIQUE",
    "kind": "function",
    "label": "UNIQUE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "UPDATE",
    "kind": "function",
    "label": "UPDATE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "VALUES",
    "kind": "function",
    "label": "VALUES",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "VIEW",
    "kind": "function",
    "label": "VIEW",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
  Object {
    "insertText": "WHERE",
    "kind": "function",
    "label": "WHERE",
    "range": Object {
      "endColumn": 1,
      "endLineNumber": 1,
      "startColumn": 1,
      "startLineNumber": 1,
    },
  },
]
`;
